### ISSUE TITLE
Flow refinement is ignored when initializing the store for inner classes

### ISSUE THREAD

**Issue by GoogleCodeExporter** [typetools/checker-framework#266]
List of test cases:
checker/tests/nullness/FinalVar2.java
checker/tests/nullness/Issue411.java
checker/tests/nullness/Issue811java
checker/tests/nullness/Issue266.java
checker/tests/nullness/Issue266a.java

```
# What steps will reproduce the problem?

Run the nullness checker on the attached file.

# What is the expected output? What do you see instead?

I expected the nullness checker to recognize that the deference of 'tmp' in the 
inner class will always succeed. Instead, I got the error:

FinalNullable.java:16: error: dereference of possibly-null reference tmp
        return tmp.toString();
               ^

#What version of the product are you using? On what operating system?

javac 1.8.0-jsr308-1.6.7 (7102:18502dce6db5)
```

Original issue reported on code.google.com by `cus...@google.com` on 16 Oct 2013 at 4:45

Attachments:
- [FinalNullable.java](https://storage.googleapis.com/google-code-attachments/checker-framework/issue-266/comment-0/FinalNullable.java)

**Comment by scottkennedy** [comment]
While the workaround seems reasonable, it would be nice to not require to change valid code to meet the requirements of a static analysis tool.

**Comment by smillst** [comment]
This issue was reported by another user in #811.

**Comment by smillst** [comment]
I added a test case checker/tests/nullness/Issue811java.

**Comment by wmdietl** [comment]
`org.checkerframework.dataflow.cfg.CFGBuilder.CFGTranslationPhaseOne.visitNewClass(NewClassTree, Void)` contains the comment "We ignore any class body because its methods should be visited separately.".
This is likely the location where the current store should be propagated to that anonymous class.

Handling of inner/anonymous classes probably needs a solution similar to how lambda expressions are handled. Note how rewriting an anonymous class to an equivalent lambda expression gives the correct refinement behavior.
See `org.checkerframework.dataflow.cfg.CFGBuilder.declaredLambdas` and `org.checkerframework.framework.type.GenericAnnotatedTypeFactory.analyze(Queue<ClassTree>, Queue<Pair<LambdaExpressionTree, Store>>, UnderlyingAST, List<Pair<VariableElement, Value>>, ClassTree, boolean, boolean, boolean, Store)`

**Comment by mihai-dumitresq** [comment]
Is there a cleaner annotation based workaround rather the local variable copy? (I'm thinking something like @MonotonicNonNull - although this one specifically doesn't work).

**Comment by wmdietl** [comment]
I don't think an approach with an annotation on the field is possible. Annotating the field would not help you with determining whether the type was refined or not. Flow-sensitivity needs to be implemented correctly, which is a considerable engineering effort.
I wrote up a few comments last week, but won't get to work on this for a while.

**Comment by mihai-dumitresq** [comment]
Issue #903 was closed as a dupe of this. Given that, can you add the following as a test when fixing this issue? Thanks!

```Java
class Foo {
  private final Object mBar;

  public Foo() {
    mBar = "test";
    Runnable runnable = new Runnable() {
        @Override
        public void run() {
          // error: [dereference.of.nullable] dereference of possibly-null reference mBar
          mBar.toString();
        }
      };
    runnable.run();
  }
}
```

**Comment by mernst** [comment]
Thanks for the test case.  In commit ce3465d, I have added two test cases as files checker/tests/nullness/Issue266.java and checker/tests/nullness/Issue266a.java.

**Comment by wmdietl** [comment]
Fixed by https://github.com/typetools/checker-framework/commit/c74edd7884c7682f37ad49db401de84458023c66
